src/value/borrowed.rs

```diff
@@ -0,0 +1,229 @@
+    pub fn new(table: &'val str, sid: SymbolId) -> Self {
+        Self { table, sid }
+    }
+}
+
+impl<'val> ImportSource for BorrowedImportSource<'val> {
+    fn table(&self) -> &str {
+        self.table
+    }
+
+    fn sid(&self) -> usize {
+        self.sid
+    }
+}
+
+/// A borrowed implementation of [`SymbolToken`].
+#[derive(Debug, Copy, Clone)]
+pub struct BorrowedSymbolToken<'val> {
+    text: Option<&'val str>,
+    local_sid: Option<SymbolId>,
+    source: Option<BorrowedImportSource<'val>>,
+}
+
+impl<'val> BorrowedSymbolToken<'val> {
+    pub fn new(
+        text: Option<&'val str>,
+        local_sid: Option<SymbolId>,
+        source: Option<BorrowedImportSource<'val>>,
+    ) -> Self {
+        Self {
+            text,
+            local_sid,
+            source,
+        }
+    }
+}
+
+impl<'val> From<&'val str> for BorrowedSymbolToken<'val> {
+    fn from(text: &'val str) -> Self {
+        Self::new(Some(text), None, None)
+    }
+}
+
+impl<'val> SymbolToken for BorrowedSymbolToken<'val> {
+    type ImportSource = BorrowedImportSource<'val>;
+
+    fn text(&self) -> Option<&str> {
+        self.text
+    }
+
+    fn local_sid(&self) -> Option<usize> {
+        self.local_sid
+    }
+
+    fn source(&self) -> Option<&Self::ImportSource> {
+        self.source.as_ref()
+    }
+}
+
+/// A borrowed implementation of [`Sequence`].
+#[derive(Debug, Clone)]
+pub struct BorrowedSequence<'val> {
+    children: Vec<BorrowedElement<'val>>,
+}
+
+impl<'val> BorrowedSequence<'val> {
+    pub fn new(children: Vec<BorrowedElement<'val>>) -> Self {
+        Self { children }
+    }
+}
+
+impl<'val> Sequence for BorrowedSequence<'val> {
+    type Element = BorrowedElement<'val>;
+
+    fn iter<'a>(&'a self) -> Box<dyn Iterator<Item = &'a Self::Element> + 'a> {
+        Box::new(self.children.iter())
+    }
+}
+
+/// A borrowed implementation of [`Struct`]
+#[derive(Debug, Clone)]
+pub struct BorrowedStruct<'val> {
+    // TODO model actual map indexing for the struct (maybe as a variant type)
+    //      otherwise struct field lookup will be O(N)
+    fields: Vec<(BorrowedSymbolToken<'val>, BorrowedElement<'val>)>,
+}
+
+impl<'val> BorrowedStruct<'val> {
+    pub fn new(fields: Vec<(BorrowedSymbolToken<'val>, BorrowedElement<'val>)>) -> Self {
+        Self { fields }
+    }
+}
+
+impl<'val> Struct for BorrowedStruct<'val> {
+    type FieldName = BorrowedSymbolToken<'val>;
+    type Element = BorrowedElement<'val>;
+
+    fn iter<'a>(
+        &'a self,
+    ) -> Box<dyn Iterator<Item = (&'a Self::FieldName, &'a Self::Element)> + 'a> {
+        // convert &(k, v) -> (&k, &v)
+        Box::new(self.fields.iter().map(|kv| match &kv {
+            (k, v) => (k, v),
+        }))
+    }
+}
+
+// TODO replace the references with `Cow` and bridge to the owned APIs for mutability
+
+/// Variants for all borrowed version _values_ within an [`Element`].
+#[derive(Debug, Clone)]
+pub enum BorrowedValue<'val> {
```

Comment:
Do you imagine having `ToOwned`/`Borrow`/`AsRef` _etc_ for these `enum`s?

(I'm curious about how you envision moving between `BorrowedT`/`T` underlying data, or if that's even necessary; maybe all interaction needs to happen behind the trait and you get what you get)

Reply:
Yes, I want those things, but didn't factor them in this PR since I wanted to make sure it could be even abstracted with this factoring.